# Makes sure executables generated during Ruby's build process are linked against libruby-static instead of libruby on Windows.
# On Windows, linking executables against libruby instead of libruby-static results in build errors if Ruby is configured with `--with-static-linked-ext`.

--- a/configure.ac
+++ b/configure.ac
@@ -3808,8 +3808,8 @@ AS_CASE(["$target_os"],
 	    LIBRUBY='lib$(RUBY_SO_NAME).dll.a'
 	], [
 	    LIBRUBY_SO=dummy
-	    LIBRUBY='lib$(RUBY_SO_NAME).a'
-	    LIBRUBYARG='-l$(RUBY_SO_NAME)'
+	    LIBRUBY='lib$(RUBY_SO_NAME)-static.a'
+	    LIBRUBYARG='-l$(RUBY_SO_NAME)-implib'; EXTDLDFLAGS="${EXTDLDFLAGS} -l${RUBY_SO_NAME}-implib"; AC_SUBST(MKXP_EXE_NAME)
 	])
 	],
     [hpux*], [
--- a/cygwin/GNUmakefile.in
+++ b/cygwin/GNUmakefile.in
@@ -22,8 +22,8 @@ else
 endif
 
 ifneq ($(ENABLE_SHARED),yes)
-  RUBY_EXP = $(RUBY_INSTALL_NAME).exp
-  EXTOBJS = $(RUBY_EXP)
+  LIBRUBY_IMPLIB = lib$(RUBY_SO_NAME)-implib.@LIBEXT@
+
   LIBRUBYARG = $(LIBRUBY_A)
   LIBRUBY_SO =
 endif
@@ -46,7 +46,7 @@ RUBYDEF = $(DLL_BASE_NAME).def
 ruby: $(PROGRAM)
 rubyw: $(WPROGRAM)
 
-$(LIBRUBY): $(RUBY_EXP) $(LIBRUBY_SO)
+$(LIBRUBY): $(RUBY_EXP) $(LIBRUBY_SO) $(LIBRUBY_IMPLIB)
 $(RUBY_EXP) $(LIBRUBY_SO): $(DLL_BASE_NAME).res.@OBJEXT@
 
 %.res.@OBJEXT@: %.rc
@@ -102,3 +102,8 @@ clean-local::
 	@$(RM) $(RUBYDEF)
 	@$(RM) $(RUBY_EXP) $(RCFILES:.rc=.res.@OBJEXT@)
 	@$(RM) $(RCFILES)
+
+$(LIBRUBY_IMPLIB): $(OBJS) dmyext.o dmyenc.o
+	mkdir $@.p
+	$(LDSHARED) $(DLDFLAGS) -Wl,--export-all-symbols,--out-implib,$@,--output-def,$@.def -o $@.p/@MKXP_EXE_NAME@@EXEEXT@ $^ $(LIBS)
+	rm -rf $@.p
